src/bin/ion/commands/beta/generate/generator.rs

```diff
@@ -21,6 +21,8 @@ pub(crate) struct CodeGenerator<'a, L: Language> {
     output: &'a Path,
     // Represents a counter for naming anonymous type definitions
     pub(crate) anonymous_type_counter: usize,
+    // Current type definition is root type or not
+    is_root_type: bool,
```

Comment:
Please expand on this comment. What is the significance of knowing whether or not this is the root type?

Reply:
For example, we have the following type definition,
```
type::{
  name: struct_type,
  fields: {
    foo: foo_type
  }
}
type::{
 name: foo_type,
 type: int
}
```
The generated code for `foo_type` will have a read API that iterates over the fields bycalling `reader.next()`, matching with the `field_name` to call appropriate `read_from` call. (Here that would be `Foo::read_from(reader)`). but when since even the nested types use the same templates the first statement inside this `read_from` API is calling `reader.next()`. This would skip the `foo` all together leading to end of struct. Which is why I am using `is_root_type` property make the approriate call to `reader.ext()`.

Reference template:
```rust
pub fn read_from(reader: &mut Reader) -> IonResult<Self> {
        {% if is_root_type %}
        reader.next()?;
        {% endif %}
        ...
 }
```

Reply:
This will be changed once we have lazy reader APIs which has APIs like `expect_next`.

Reply:
Have you considered changing the contract of `read_from` to say that the reader must already be positioned on the value that is being read as this type? That seems like the simpler, and more generalizable solution. (After all, how do you know to read this type if you haven't seen anything about the next value?)

Reply:
Yes, that makes sense. I used this approach so that there is no additional work (calling `reader.next()`) that user has to do on the reader before calling `read_from`. 
I was also going to use this `is_root_type` to generate tests for generated code, but with the new strategy that we discussed for testing might not require that now . Removing it for now, if required for tetsing will add later.